### DISCUSSION TITLE
Windows builds

### DISCUSSION THREAD

**Discussion by jrm523** [liquidctl/liquidctl#391] [General]
I came back and saw that the team has stopped supplying windows builds. While I am hugely disappointed. I appreciate the fact that free programs like this are available due to hard working individuals and teams. I just want to say that I will deeply miss having your program for my windows builds. Thank you very much for all of the hard work that you have performed with creating an amazing app such as this. Maybe one day I will see you on the linux side of things. However, I will still be Windows for some time.

**Comment by jonasmalacofilho** [comment]
While we're no longer publishing executable files for Windows, that is still a platform we support.

You can follow the documentation to install it from PyPI, or directly from the sources in this repository.

---

Related: #370

**Reply by Maid3n** [reply]
Hello,
I have tried to read and understand how to install and use PyPI but I guess my english is not strong enough...
If someone could share a windows compiled liquidctl v1.8.0 would be nice.
Windows Defender is not running on my PC and I have no detection problem with Eset Nod32

**Reply by jonasmalacofilho** [reply]
Can I help you with the PyPI installation?

Have you installed Python, and when doing it have you checked the option to add it to your `PATH`?

If so, you should be able to simply open a Command Prompt (or PowerShell) window and type:

```
pip install liquidctl
```

**Reply by Maid3n** [reply]
Thanks for reply.
After more investigations, a problem with pip command not recognized (solved), a problem with usb.core.NoBackendError (solved too !), liquidctl is working again as it should !

**Reply by Sabrehawk** [reply]
thanks that was the information i needed to continue my installation... i would humbly suggest adding this step to the readme for windows installations.

**Comment by Maid3n** [comment]
I have a little question.
Running Win10 x64 and using a Z63.
Is it normal that liquidctl v1.8.0 doesn't work with WinUsb driver installed? It seems to run only with libusb-win32 or libusbK installed via Zadig...
In Zadig, kraken appears twice : interface 1 with HidUsb driver, left as it, and interface 0 with WinUsb driver I had to replace.

**Reply by jonasmalacofilho** [reply]
Can you open a bug report issue with the error output from liquidctl with the WinUsb driver (on interface 0; interface 1 really must use the HID driver)?

If you can, set the `PYUSB_DEBUG=debug` and `LIBUSB_DEBUG=4` environments variables first.  And also pass `--debug` in additional to any liquidctl commands/options you're using.

**Comment by Maid3n** [comment]
Hello, how do set these variables and restore them to default values ?
For information, liquidctl.exe version 1.7.1 runs fine with standard WinUSB driver (v10) from win10 or from Zadig (v6).

here is an example of error, text copied from powershell command:

```
PS C:\WINDOWS\system32> liquidctl status --debug
[DEBUG] (cli) (main): version: liquidctl v1.8.0
[DEBUG] (cli) (main): platform: Windows-10-10.0.19043-SP0
[DEBUG] (usb) (find_devices): searching HidapiBus (CommanderCore, CommanderPro, CorsairHidPsu, HydroPlatinum, Kraken2, KrakenX3, KrakenZ3, NzxtEPsu, RgbFusion2, SmartDevice, SmartDevice2, _CommonSmartDeviceDriver)
[DEBUG] (usb) (find_devices): found HID device 1e7d:2e23
[DEBUG] (usb) (find_devices): found HID device 1e7d:2e23
[DEBUG] (usb) (find_devices): found HID device 1e7d:2e23
[DEBUG] (usb) (find_devices): found HID device 1e7d:2e23
[DEBUG] (usb) (find_devices): found HID device 1e7d:2e23
[DEBUG] (usb) (find_devices): found HID device 1b1c:1b09
[DEBUG] (usb) (find_devices): found HID device 1b1c:1b09
[DEBUG] (usb) (find_devices): found HID device 1b1c:1b09
[DEBUG] (usb) (find_devices): found HID device 1e71:3008
[DEBUG] (usb) (probe): instantiated KrakenZ3 driver for NZXT Kraken Z (Z53, Z63 or Z73)
[DEBUG] (usb) (find_devices): searching PyUsbBus (Hydro690Lc, Legacy690Lc, Modern690Lc, _Base690Lc, _ModernBase690Lc, CorsairAsetekProDriver)
Traceback (most recent call last):
  File "C:\Python\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Python\Scripts\liquidctl.exe\__main__.py", line 7, in <module>
  File "C:\Python\lib\site-packages\liquidctl\cli.py", line 373, in main
    selected = list(find_liquidctl_devices(**opts))
  File "C:\Python\lib\site-packages\liquidctl\driver\__init__.py", line 59, in find_liquidctl_devices
    for dev in bus_cls().find_devices(**kwargs):
  File "C:\Python\lib\site-packages\liquidctl\driver\usb.py", line 547, in find_devices
    for handle in PyUsbDevice.enumerate(vendor, product):
  File "C:\Python\lib\site-packages\liquidctl\driver\usb.py", line 320, in enumerate
    for handle in usb.core.find(find_all=True, **args):
  File "C:\Python\lib\site-packages\usb\core.py", line 1309, in find
    raise NoBackendError('No backend available')
usb.core.NoBackendError: No backend available
```


My kraken device is 1e71:3008

Here is one of my batch file that works with liquidctlexe v1.7.1 + WinUsb driver and with liquidctl v1.8.0 + libusb driver...

```bat
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo - - - -                                       - - - -
@echo - - - -    N Z X T   K R A K E N    Z  6 3    - - - -
@echo - - - -                                       - - - -
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo - - - -                                       - - - -
@echo - - - -        M O D E   S I L E N C E        - - - -
@echo - - - -                                       - - - -
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo - - - - - - - - - - - - - - - - - - - - - - - - - - -
@echo off
timeout /t 3 /nobreak > NUL
liquidctl initialize
rem liquidctl set sync color off (not working with Z model)
liquidctl set external color off
liquidctl set fan speed 20 20  30 30  40 40  45 60  50 70  55 80
liquidctl set pump speed 20 50  30 50  40 70  55 100
timeout /t 3 /nobreak > NUL
```

Any liquidctl command returns same  debug error text.

**Reply by jonasmalacofilho** [reply]
> Hello, how do set these variables and restore them to default values ?

I can never remember, since I don't use Windows that much (except for gaming), and even then it's been more than a month.

I think on Command Prompt you need to use `set`:

```
> set PYUSB_DEBUG=debug
> set LIBUSB_DEBUG=4
> liquidctl ...
```

PowerShell seems to work [differently](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.2).

In both cases I think you need to set the variables first, then run `liquidctl`.  But afterwards you don't really need to restore the default values, since the changes will be local to the Command Prompt/PowerShell session.

---

The `usb.core.NoBackendError` suggests we're simply not finding the `libusb-1.0.dll`.

> The libusb DLLs can be found in [libusb/releases](https://github.com/libusb/libusb/releases) (part of the `libusb-<version>.7z` files) and the appropriate (e.g. MS64) `.dll` and `.lib` files should be extracted to the system or python installation directory (e.g. `C:\Windows\System32` or `C:\Python36`).[^1]

Placing them on the system folder is usually easier and more reliable.

[^1]: From the README: Installing on Windows

---

> For information, liquidctl.exe version 1.7.1 runs fine with standard WinUSB driver (v10) from win10 or from Zadig (v6).

Where you using the .exe?  That could explain the different behavior, since the .exe has its own bundled copy of `libusb-1.0.dll`.

**Reply by Maid3n** [reply]
Yes, I was using the 1.7.1 exe

**Comment by Maid3n** [comment]
here is the list of package installed in python:

```
PS C:\WINDOWS\system32> pip list
Package             Version
------------------- ------------
colorama            0.4.4
colorlog            6.6.0
docopt              0.6.2
hidapi              0.11.0.post2
importlib-metadata  4.10.0
importlib-resources 5.4.0
libusb              1.0.24b3
liquidctl           1.8.0
packaging           21.3
pip                 21.3.1
pkg-about           1.0.4
pyparsing           3.0.6
pyusb               1.2.1
setuptools          57.4.0
zipp                3.7.0
```

**Reply by jonasmalacofilho** [reply]
> ```
> libusb              1.0.24b3
> ```

I don't the Python `libusb` package installs its `libusb-1.0.dll` in a way that's immediately accessible to the system.

(And PyUSB simply looks for libusb on the system, which makes more sense for Linux and, arguably, even Mac).

**Comment by Maid3n** [comment]
I've extracted dll from MinGW64 folder to my windows\system32 folder and everything is ok now with liquidctl v1.8.0.

Thanks a lot for your help, !

**Reply by jonasmalacofilho** [reply]
Thanks for letting me know!

**Comment by Maid3n** [comment]
I'm very happy with this little app, CAM is not needed anymore and that's great !

**Comment by chimcen** [comment]
Hi, sry for this potential stupid question - but can you share a how to for compiling liquidctl to an executable in windows environment? I've allready installed it in a python venv and it's working, but it would be nice to get an executable out of it to share it with friends...

**Reply by jonasmalacofilho** [reply]
What if you became responsible for getting executable past antivirus false-positives (see #370)?

This way we would avoid having unofficial executables being distributed in the wild, and at the same improve the easy of setup for all Windows users.

**Reply by chimcen** [reply]
thats not a problem for me - because i'll share it only with my brother and a friend... if they got av problems they have to deal their self with it ;)

**Reply by jonasmalacofilho** [reply]
Have a go at the artifacts from our [AppVeyor CI builds](https://ci.appveyor.com/project/jonasmalacofilho/liquidctl/history).

Note: since #370, the executables built there aren't being maintained or tested on real hardware.

**Reply by jonasmalacofilho** [reply]
@chimcen,

Just a heads up: the .exe build on AppVeyor started to mysteriously fail, and I disabled it to avoid blocking our tests from running:

- [AppVeyor build #1571](https://ci.appveyor.com/project/jonasmalacofilho/liquidctl/builds/42467666)
- Commit liquidctl/liquidctl@9fc2ff1083aa169cce7125309f0ee66fdf214175 ("appveyor: disable liquidctl.exe builds with pyinstaller")

I know you're interested in those executables, so please consider trying to fix the issue on AppVeyor, or maybe even switching the Windows tests & build to GitHub Actions (where we already run all non-Windows tests).
